File: producer_consumer.py
"""
Producer-Consumer Pattern
Demonstrates the classic producer-consumer pattern using threading and queues.
"""
import threading
import queue
import time
import random


def producer(queue_obj, producer_id, num_items):
    """
    Producer function that generates items and puts them in the queue.
    
    Args:
        queue_obj: Queue to put items into
        producer_id: Unique identifier for the producer
        num_items: Number of items to produce
    """
    for i in range(num_items):
        item = f"Item-{producer_id}-{i}"
        queue_obj.put(item)
        print(f"Producer {producer_id} produced: {item}")
        time.sleep(random.uniform(0.1, 0.5))
    print(f"Producer {producer_id} finished")


def consumer(queue_obj, consumer_id, stop_event):
    """
    Consumer function that processes items from the queue.
    
    Args:
        queue_obj: Queue to get items from
        consumer_id: Unique identifier for the consumer
        stop_event: Event to signal when to stop
    """
    while not stop_event.is_set() or not queue_obj.empty():
        try:
            item = queue_obj.get(timeout=0.5)
            print(f"Consumer {consumer_id} consumed: {item}")
            time.sleep(random.uniform(0.1, 0.3))
            queue_obj.task_done()
        except queue.Empty:
            continue
    print(f"Consumer {consumer_id} finished")


def run_producer_consumer_example():
    """Run producer-consumer pattern example."""
    print("=== Producer-Consumer Pattern Example ===\n")
    
    # Create a queue with max size
    work_queue = queue.Queue(maxsize=10)
    stop_event = threading.Event()
    
    # Create producer threads
    num_producers = 2
    num_items_per_producer = 5
    producers = []
    for i in range(num_producers):
        thread = threading.Thread(target=producer, args=(work_queue, i, num_items_per_producer))
        producers.append(thread)
        thread.start()
    
    # Create consumer threads
    num_consumers = 3
    consumers = []
    for i in range(num_consumers):
        thread = threading.Thread(target=consumer, args=(work_queue, i, stop_event))
        consumers.append(thread)
        thread.start()
    
    # Wait for all producers to finish
    for thread in producers:
        thread.join()
    
    # Wait for queue to be empty
    work_queue.join()
    
    # Signal consumers to stop
    stop_event.set()
    
    # Wait for all consumers to finish
    for thread in consumers:
        thread.join()
    
    print("\nAll producers and consumers finished!\n")


if __name__ == "__main__":
    run_producer_consumer_example()

File: src/main/java/com/multithreading/ProducerConsumer.java
package com.multithreading;

import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.Random;

/**
 * Producer-Consumer Pattern
 * Demonstrates the classic producer-consumer pattern using BlockingQueue.
 */
public class ProducerConsumer {

    static class Producer extends Thread {
        private final BlockingQueue<String> queue;
        private final int producerId;
        private final int numItems;

        public Producer(BlockingQueue<String> queue, int producerId, int numItems) {
            this.queue = queue;
            this.producerId = producerId;
            this.numItems = numItems;
        }

        @Override
        public void run() {
            Random random = new Random();
            for (int i = 0; i < numItems; i++) {
                try {
                    String item = "Item-" + producerId + "-" + i;
                    queue.put(item);
                    System.out.println("Producer " + producerId + " produced: " + item);
                    Thread.sleep(random.nextInt(500) + 100);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    break;
                }
            }
            System.out.println("Producer " + producerId + " finished");
        }
    }

    static class Consumer extends Thread {
        private final BlockingQueue<String> queue;
        private final int consumerId;
        private volatile boolean running = true;

        public Consumer(BlockingQueue<String> queue, int consumerId) {
            this.queue = queue;
            this.consumerId = consumerId;
        }

        public void stopConsuming() {
            running = false;
        }

        @Override
        public void run() {
            Random random = new Random();
            while (running || !queue.isEmpty()) {
                try {
                    String item = queue.poll(500, java.util.concurrent.TimeUnit.MILLISECONDS);
                    if (item != null) {
                        System.out.println("Consumer " + consumerId + " consumed: " + item);
                        Thread.sleep(random.nextInt(300) + 100);
                    }
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    break;
                }
            }
            System.out.println("Consumer " + consumerId + " finished");
        }
    }

    public static void runExample() throws InterruptedException {
        System.out.println("=== Producer-Consumer Pattern Example ===\n");

        BlockingQueue<String> queue = new LinkedBlockingQueue<>(10);

        // Create and start producers
        int numProducers = 2;
        int numItemsPerProducer = 5;
        Producer[] producers = new Producer[numProducers];
        for (int i = 0; i < numProducers; i++) {
            producers[i] = new Producer(queue, i, numItemsPerProducer);
            producers[i].start();
        }

        // Create and start consumers
        int numConsumers = 3;
        Consumer[] consumers = new Consumer[numConsumers];
        for (int i = 0; i < numConsumers; i++) {
            consumers[i] = new Consumer(queue, i);
            consumers[i].start();
        }

        // Wait for all producers to finish
        for (Producer producer : producers) {
            producer.join();
        }

        // Give consumers time to finish remaining items
        Thread.sleep(1000);

        // Stop all consumers
        for (Consumer consumer : consumers) {
            consumer.stopConsuming();
        }

        // Wait for all consumers to finish
        for (Consumer consumer : consumers) {
            consumer.join();
        }

        System.out.println("\nAll producers and consumers finished!\n");
    }

    public static void main(String[] args) throws InterruptedException {
        runExample();
    }
}
